lib/si-id/src/macros.rs

```diff
@@ -0,0 +1,299 @@
+        );
+
+        impl Default for $name {
+            fn default() -> Self {
+                Self::NONE
+            }
+        }
+
+        impl $name {
+            /// The "nil" value for this ID.
+            pub const NONE: Self = Self(::ulid::Ulid::nil());
+        }
+    };
+}
+
+/// **Deprecated:** this macro provides additional functionality for using the "nil" ID on top of the standard `id!`
+/// macro. It also adds support for PostgreSQL type conversions.
+///
+/// This absolutely should not be used for any new IDs. Nick will be mad. No, I'm not writing this in third person, why
+/// do you ask?
+macro_rules! id_with_none_and_pg_types {
+    (
+        $(#[$($attrs:tt)*])*
+        $name:ident
+    ) => {
+        do_not_use_directly_id_inner!(
+            $(#[$($attrs)*])*
+            $name
+        );
+
+        impl Default for $name {
+            fn default() -> Self {
+                Self::NONE
+            }
+        }
+
+        impl $name {
+            /// The "nil" value for this ID.
+            pub const NONE: Self = Self(::ulid::Ulid::nil());
+        }
+
+        impl<'a> postgres_types::FromSql<'a> for $name {
+            fn from_sql(
+                ty: &postgres_types::Type,
+                raw: &'a [u8],
+            ) -> Result<Self, Box<dyn std::error::Error + Sync + Send>> {
+                let id: String = postgres_types::FromSql::from_sql(ty, raw)?;
+                Ok(Self(::ulid::Ulid::from_string(&id)?))
+            }
+
+            fn accepts(ty: &postgres_types::Type) -> bool {
+                ty == &postgres_types::Type::BPCHAR
+                    || ty.kind() == &postgres_types::Kind::Domain(postgres_types::Type::BPCHAR)
+            }
+        }
+
+        impl postgres_types::ToSql for $name {
+            fn to_sql(
+                &self,
+                ty: &postgres_types::Type,
+                out: &mut postgres_types::private::BytesMut,
+            ) -> Result<postgres_types::IsNull, Box<dyn std::error::Error + Sync + Send>>
+            where
+                Self: Sized,
+            {
+                postgres_types::ToSql::to_sql(&self.0.to_string(), ty, out)
+            }
+
+            fn accepts(ty: &postgres_types::Type) -> bool
+            where
+                Self: Sized,
+            {
+                ty == &postgres_types::Type::BPCHAR
+                    || ty.kind() == &postgres_types::Kind::Domain(postgres_types::Type::BPCHAR)
+            }
+
+            fn to_sql_checked(
+                &self,
+                ty: &postgres_types::Type,
+                out: &mut postgres_types::private::BytesMut,
+            ) -> Result<postgres_types::IsNull, Box<dyn std::error::Error + Sync + Send>> {
+                postgres_types::ToSql::to_sql(&self.0.to_string(), ty, out)
+            }
+        }
+    };
+}
+
+// TODO(nick): figure out how to scope this for only macros in this module.
+#[allow(missing_docs)]
+macro_rules! do_not_use_directly_id_inner {
```

Comment:
Open to ideas on how to ensure this macro isn't exposed outside of this _file_. Low priority for this PR though.

Reply:
This isn't possible in that way that we want it with stable Rust right now. Amended the comment accordingly